src/FTThresholdClient.cpp

```diff
@@ -0,0 +1,62 @@
+#include "rewd_controllers/FTThresholdClient.hpp"
+#include <thread>
+
+namespace rewd_controllers {
+
+using SetFTThresholdAction = pr_control_msgs::SetForceTorqueThresholdAction;
+using FTThresholdActionClient
+    = actionlib::SimpleActionClient<SetFTThresholdAction>;
+
+//=============================================================================
+FTThresholdClient::FTThresholdClient(const std::string& controllerThresholdTopic)
+{
+  mFTThresholdActionClient = std::unique_ptr<FTThresholdActionClient>(
+      new FTThresholdActionClient(controllerThresholdTopic));
+  ROS_INFO("Waiting for FT Threshold Action Server to start...");
+  mFTThresholdActionClient->waitForServer();
+  ROS_INFO("FT Threshold Action Server started.");
+}
+
+//=============================================================================
+bool FTThresholdClient::trySetThresholdRepeatedly(double forceThreshold, double torqueThreshold)
+{
+  while (ros::ok())
+  {
+    try {
+    if (setThreshold(forceThreshold, torqueThreshold))
+      return true;
+    } catch (std::runtime_error) {}
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+  return false;
+}
+
+//=============================================================================
+bool FTThresholdClient::setThreshold(double forceThreshold, double torqueThreshold, double timeout)
+{
+  pr_control_msgs::SetForceTorqueThresholdGoal goal;
+  goal.force_threshold = forceThreshold;
+  goal.torque_threshold = torqueThreshold;
+  mFTThresholdActionClient->sendGoal(goal);
+  bool finished_before_timeout
+      = mFTThresholdActionClient->waitForResult(ros::Duration(timeout));
+
+  if (finished_before_timeout)
+  {
+    actionlib::SimpleClientGoalState state
+        = mFTThresholdActionClient->getState();
+    if (state != actionlib::SimpleClientGoalState::StateEnum::SUCCEEDED)
```

Comment:
It seems like part of the problem is that there's no way for the client to distinguish whether the request failed because (1) the controller hasn't started, (2) taring isn't complete, or (3) the thresholds are bad (negative or larger than the limit). The server is marking all these states as `REJECTED`, so all these errors are getting squashed into one `runtime_error`. This means that we could possibly wind up in an infinite loop if we send bad thresholds, which isn't ideal.

Maybe one way to distinguish the two would be to use `REJECTED` and `ABORTED`. I'm still not entirely clear what the convention is in actionlib (the [wiki](http://wiki.ros.org/actionlib/DetailedDescription#Server_Description) is only somewhat helpful) but maybe we can use `REJECTED` for cases (1) and (3) when the user is doing something _really_ wrong and `ABORTED` for case (2) when the user just needs to wait for taring to finish.

Then on the client side, we could raise an error in `setThresholds` for `REJECTED` and just return `false` for `ABORTED`. That would make the try-catch in the retry version unnecessary, since a `REJECTED` request will never succeed after retrying.

@gilwoolee what do you think?

Reply:
I think that's a good idea. So with regards to trajectory goals we actually had a small discussion, about the meaning of these terms, where my belief is that `CANCELED` means that the user stopped the trajectory, while `ABORTED` means that the controller stopped the trajectory for a reason that does not constitute a failure (like when the `MoveUntilTouch` touches something).

Now, of course the action server doesn't have `CANCELED`, but I think the meaning of the `ABORTED` action matches pretty well with your description, so I think that's a good way to handle the `setThresholds`, too.

Reply:
I'm in favor of `ABORTED`, just so we have more information without having to dig into the error message.

Reply:
Alright, I think we have one last decision left:
After the controller `ABORTED` it, the `FTThresholdClient` can return `false` or throw an exception.
In the commit I just pushed, it throws.

Reply:
I think `ABORTED` should result in `false` rather than an exception. So this condition would become
```cpp
if (state == actionlib::SimpleClientGoalState::StateEnum::REJECTED)
  throw std::runtime_error(...);
if (state == actionlib::SimpleClientGoalState::StateEnum::ABORTED)
  return false;
if (state == actionlib::SimpleClientGoalState::StateEnum::SUCCEEDED)
  return true;
```

Then, `trySetThresholdsRepeatedly` can simply call `setThresholds` in the loop without the try-catch.